File: internal/analyzer/detectors/unusual_ttl.py
from typing import Dict, List, Set

import pandas as pd

from internal.analyzer.base import Alert, BaseDetector, Level


class UnusualTTLDetector(BaseDetector):
    def __init__(self):
        super().__init__()
        # Standard OS values: 64 (Linux/Android), 128 (Windows), 255 (Network equipment)
        self.standard_ttls = {64, 128, 255}

    def analyze(self, df: pd.DataFrame) -> List[Alert]:
        alerts = []
        if df.empty or "src_ip" not in df.columns or "ttl" not in df.columns:
            return alerts

        clean_df = df.dropna(subset=["src_ip", "ttl"])
        clean_df = clean_df[
            (clean_df["src_ip"].str.strip() != "") & (clean_df["ttl"] != 0)
        ]

        if clean_df.empty:
            return alerts

        for _, row in clean_df.iterrows():
            ip = row["src_ip"]
            ttl = int(row["ttl"])

            # Ignore standard OS values
            if ttl in self.standard_ttls:
                continue

            is_local = ip.startswith(("192.168.", "10.", "172.16."))

            if is_local:
                # Ignore TTL=1 and TTL=2 for local traffic (standard for discovery)
                if ttl in {1, 2}:
                    continue
                # For local IPs, we follow the instruction to "Only alert on 'unusual' values from external IPs"
                continue
            else:
                # External IP: Only alert on "unusual" values (e.g., between 10 and 30)
                if 10 <= ttl <= 30:
                    if self._should_alert("UNUSUAL_TTL", ip):
                        alerts.append(
                            Alert(
                                timestamp=str(row["timestamp"]),
                                level=Level.INFO,
                                type="UNUSUAL_TTL",
                                message=f"External IP {ip} with unusual TTL: {ttl}",
                                src_ip=ip,
                            )
                        )

        return alerts

File: internal/analyzer/detectors/unusual_ttl.go
package detectors

import (
	"fmt"
	"hattic/internal/analyzer"
	"strings"
	"time"
)

type UnusualTTLDetector struct {
	*analyzer.BaseDetector
	standardTTLs map[uint8]bool
}

func NewUnusualTTLDetector() *UnusualTTLDetector {
	return &UnusualTTLDetector{
		BaseDetector: analyzer.NewBaseDetector(5 * time.Minute),
		standardTTLs: map[uint8]bool{
			64:  true,
			128: true,
			255: true,
		},
	}
}

func (d *UnusualTTLDetector) Analyze(packet *analyzer.PacketInfo) []analyzer.Alert {
	var alerts []analyzer.Alert

	if packet.SrcIP == "" || packet.TTL == 0 {
		return alerts
	}

	if d.standardTTLs[packet.TTL] {
		return alerts
	}

	isLocal := strings.HasPrefix(packet.SrcIP, "192.168.") ||
		strings.HasPrefix(packet.SrcIP, "10.") ||
		strings.HasPrefix(packet.SrcIP, "172.16.")

	if isLocal {
		return alerts
	}

	if packet.TTL >= 10 && packet.TTL <= 30 {
		if d.ShouldAlert("UNUSUAL_TTL", packet.SrcIP) {
			alerts = append(alerts, analyzer.Alert{
				Timestamp: packet.Timestamp.Format(time.RFC3339),
				Level:     analyzer.LevelInfo,
				Type:      "UNUSUAL_TTL",
				Message:   fmt.Sprintf("External IP %s with unusual TTL: %d", packet.SrcIP, packet.TTL),
				SrcIP:     packet.SrcIP,
			})
		}
	}

	return alerts
}
